chainer/link.py

```diff
@@ -134,6 +141,46 @@ def add_param(self, name, shape, dtype=numpy.float32):
         var.grad = grad
         self._params.append(name)
         d[name] = var
+        if name in self._uninitialized_params:
+            self._uninitialized_params.remove(name)
+
+    def add_uninitialized_param(self, name):
+        """Registers an uninitialized parameter to the link.
+
+        An uninitialized parameter is defined as a parameter that has a name
+        but that does not yet have a shape. If the shape of a parameter
+        depends on the shape of the inputs to the ``__call__`` operator,
+        it can be useful to defer initialization (that is, setting the shape)
+        until the first forward call of the link. Such parameters are
+        intended to be defined as uninitialized parameters in the initializer
+        and then initialized during the first forward call.
+
+        An uninitialized parameter is intended to be registered to a link by
+        calling this method in the initializer method. Then, during the
+        first forward call, the shape of the parameter will be determined
+        from the size of the inputs and the parameter must be initialized by
+        calling the :meth:`add_param` method.
+
+        Args:
+            name: (str): Name of the uninitialized parameter.
+
+        """
+        d = self.__dict__
+        if (name in self._uninitialized_params) or (name in d):
+            raise AttributeError(
+                'cannot register a new uninitialized parameter %s: exists'
+                % name)
+        self._uninitialized_params.add(name)
+
+    @property
+    def has_uninitialized_params(self):
+        """Check if the link has uninitialized parameters.
+
+        Returns: True if the link has any uninitialized parameters. Otherwise
```

Comment:
Insert  a new line after `Returns:` and add type `bool:` before the description.